core/trino-main/src/main/java/io/trino/operator/output/BlockTypeDispatchingPositionsAppender.java

```diff
@@ -0,0 +1,175 @@
+ */
+package io.trino.operator.output;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import static java.util.Objects.requireNonNull;
+
+public class BlockTypeDispatchingPositionsAppender
+        implements PositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BlockTypeDispatchingPositionsAppender.class).instanceSize();
+
+    private final PositionsAppender delegate;
+
+    public BlockTypeDispatchingPositionsAppender(PositionsAppender delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block source)
+    {
+        if (positions.isEmpty()) {
+            return;
+        }
+
+        if (source instanceof RunLengthEncodedBlock) {
+            delegate.appendRle(flattenRle((RunLengthEncodedBlock) source, positions.size()));
+        }
+        else if (source instanceof DictionaryBlock) {
+            appendDictionary(positions, (DictionaryBlock) source);
+        }
+        else {
+            delegate.append(positions, source);
+        }
+    }
+
+    private void appendDictionary(IntArrayList positions, DictionaryBlock source)
+    {
+        BlockView flatDictionary = flattenDictionary(source, positions);
+        if (flatDictionary.getBlock() instanceof RunLengthEncodedBlock) {
+            delegate.appendRle((RunLengthEncodedBlock) flatDictionary.getBlock());
+        }
+        else {
+            delegate.append(flatDictionary.getPositions(), flatDictionary.getBlock());
+        }
+    }
+
+    private BlockView flattenDictionary(DictionaryBlock source, IntArrayList positions)
+    {
+        Block dictionary = source.getDictionary();
+
+        while (dictionary instanceof RunLengthEncodedBlock || dictionary instanceof DictionaryBlock) {
+            if (dictionary instanceof RunLengthEncodedBlock) {
+                // since at some level, dictionary contains only a single value, it can be flattened to rle
+                RunLengthEncodedBlock rleDictionary = flattenRle((RunLengthEncodedBlock) dictionary, positions.size());
+                return new BlockView(rleDictionary, positions);
+            }
+            else {
+                // dictionary is a nested dictionary. we need to remap the ids
+                DictionaryBlock dictionaryValue = (DictionaryBlock) dictionary;
+                positions = mapPositions(positions, source);
+                dictionary = dictionaryValue.getDictionary();
+                source = dictionaryValue;
+            }
+        }
+        return new BlockView(dictionary, mapPositions(positions, source));
+    }
+
+    private RunLengthEncodedBlock flattenRle(RunLengthEncodedBlock source, int positionCount)
+    {
+        Block value = source.getValue();
```

Comment:
just use `getSingleValueBlock`. You can use check argument to make sure it's not nested